File: ui/client/src/api/attachments.ts
import axios from '@/http/axiosAgentConfig';

export interface AttachmentContent {
  filename: string;
  extension: string;
  text_content: string;
  char_count: number;
}

export interface AttachmentUploadResponse {
  attachments: AttachmentContent[];
}

export interface AttachmentValidationError {
  file_name: string;
  error_type: 'extension' | 'size';
  message: string;
}

export interface SupportedTypesResponse {
  allowed_extensions: string[];
  max_file_size_bytes: number;
  max_file_size_mb: number;
}

/**
 * Upload files and extract text content for prompt attachment.
 * Files are base64-encoded and sent to the backend for processing.
 */
export async function uploadAndProcessAttachments(
  files: { name: string; content: string }[]
): Promise<AttachmentUploadResponse> {
  const response = await axios.post<AttachmentUploadResponse>(
    '/attachments/upload-and-process',
    { files }
  );
  return response.data;
}

/**
 * Get supported file types for prompt attachments.
 */
export async function getAttachmentSupportedTypes(): Promise<SupportedTypesResponse> {
  const response = await axios.get<SupportedTypesResponse>(
    '/attachments/supported-types'
  );
  return response.data;
}

export const ATTACHMENT_ALLOWED_EXTENSIONS = ['.pdf', '.docx', '.md'];
export const ATTACHMENT_MAX_FILE_SIZE_MB = 10;
export const ATTACHMENT_MAX_FILE_SIZE_BYTES = ATTACHMENT_MAX_FILE_SIZE_MB * 1024 * 1024;

/**
 * Client-side validation of files before upload.
 */
export function validateAttachmentFiles(
  files: File[]
): { valid: File[]; errors: AttachmentValidationError[] } {
  const valid: File[] = [];
  const errors: AttachmentValidationError[] = [];

  for (const file of files) {
    const ext = '.' + file.name.split('.').pop()?.toLowerCase();

    if (!ATTACHMENT_ALLOWED_EXTENSIONS.includes(ext)) {
      errors.push({
        file_name: file.name,
        error_type: 'extension',
        message: `Unsupported file type "${ext}". Allowed: ${ATTACHMENT_ALLOWED_EXTENSIONS.join(', ')}`,
      });
    } else if (file.size > ATTACHMENT_MAX_FILE_SIZE_BYTES) {
      errors.push({
        file_name: file.name,
        error_type: 'size',
        message: `File exceeds maximum size of ${ATTACHMENT_MAX_FILE_SIZE_MB} MB`,
      });
    } else {
      valid.push(file);
    }
  }

  return { valid, errors };
}

/**
 * Convert a File to base64 data URL string.
 */
export function fileToBase64(file: File): Promise<string> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.onload = () => resolve(reader.result as string);
    reader.onerror = reject;
    reader.readAsDataURL(file);
  });
}

File: multi-agent/api/flask/endpoints/attachments.py
from flask import Blueprint, jsonify, current_app
from global_utils.helpers.apiargs import from_body
from webargs import fields

attachments_bp = Blueprint("attachments", __name__)


@attachments_bp.route("/upload-and-process", methods=["POST"])
@from_body({
    "files": fields.List(fields.Dict(), required=True),
})
def upload_and_process(files):
    """
    Upload document attachments and extract text content.
    
    Accepts base64-encoded files (PDF, DOCX, MD).
    Returns extracted text content for each file.
    
    Request body:
        files: [{name: str, content: str (base64)}]
    
    Response:
        attachments: [{filename, extension, text_content, char_count}]
    """
    try:
        svc = current_app.container.attachment_service

        # Validate
        valid, errors = svc.validate_files(
            [{"name": f["name"], "size": len(f.get("content", ""))} for f in files]
        )
        if errors:
            return jsonify({"errors": errors}), 400

        # Process
        results = svc.process_attachments(files)
        return jsonify({
            "attachments": [r.to_dict() for r in results]
        }), 200

    except Exception as e:
        return jsonify({"error": str(e)}), 500


@attachments_bp.route("/supported-types", methods=["GET"])
def get_supported_types():
    """Return allowed file extensions for prompt attachments."""
    from attachments.models import ALLOWED_EXTENSIONS, MAX_FILE_SIZE_BYTES
    return jsonify({
        "allowed_extensions": sorted(ALLOWED_EXTENSIONS),
        "max_file_size_bytes": MAX_FILE_SIZE_BYTES,
        "max_file_size_mb": MAX_FILE_SIZE_BYTES // (1024 * 1024),
    }), 200
